[GStreamer] Make downloadbuffer limits configurable

marex opened the pull request:
#### d2b242779d249bd5ccb080e50199ec38e87a2e40
<pre>
[GStreamer] Make downloadbuffer limits configurable
<a href="https://bugs.webkit.org/show_bug.cgi?id=263166">https://bugs.webkit.org/show_bug.cgi?id=263166</a>

Reviewed by Philippe Normand.

Introduce two new environment variables which control the maximum amount
of buffered data by the downloadbuffer element. This is useful in case
it is necessary to control buffering large files, way larger than the
default downloadbuffer limits. The following new variables are added:

- WPE_SHELL_MEDIA_DISK_CACHE_SIZE_BYTES ~ downloadbuffer max-size-bytes
- WPE_SHELL_MEDIA_DISK_CACHE_SIZE_NSEC ~ downloadbuffer max-size-time

Example HTML file, note esp. the &apos;preload=&quot;auto&quot;&apos; which is mandatory:
```
&lt;video src=&quot;./file.mp4&quot; autoplay muted loop preload=&quot;auto&quot;&gt;&lt;/video&gt;
```

Example invocation to cache up to 200 MiB per file and 500s per file:
```
$ WPE_SHELL_MEDIA_DISK_CACHE_SIZE_BYTES=268435456 \
  WPE_SHELL_MEDIA_DISK_CACHE_SIZE_NSEC=500000000000 \
  WEBKIT_GST_USE_PLAYBIN3=1 \
  cog <a href="http://example/index.html">http://example/index.html</a>
```

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleStreamCollectionMessage):

Canonical link: <a href="https://commits.webkit.org/269348@main">https://commits.webkit.org/269348@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/969360fadb9f807ff6e4b9ca96d84476c9b03228

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22300 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22499 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24200 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20638 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22550 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22828 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21648 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22536 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22235 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19352 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25054 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19285 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20208 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26459 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20302 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20439 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24319 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20969 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17762 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20427 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5310 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24657 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20965 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2336,7 +2336,20 @@ void MediaPlayerPrivateGStreamer::configureDownloadBuffer(GstElement* element)
+    const char* cacheSizeStr;
+
+    cacheSizeStr = std::getenv("WPE_SHELL_MEDIA_DISK_CACHE_SIZE_BYTES");
+    if (cacheSizeStr)
```
Something like `auto cacheSizeBytes = parseInteger<unsigned long>(cacheSizeStr).value_or(100 * KB)`

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2336,7 +2336,20 @@ void MediaPlayerPrivateGStreamer::configureDownloadBuffer(GstElement* element)
+        cacheSizeNSec = (unsigned long long)strtoull(cacheSizeStr, nullptr, 10);
+
+    g_object_set(element, "max-size-bytes", cacheSizeBytes, nullptr);
+    g_object_set(element, "max-size-time", cacheSizeNSec, nullptr);
```
This can be done with a single `g_object_set()` call

Commit:
[GStreamer] Make downloadbuffer limits configurable
https://bugs.webkit.org/show_bug.cgi?id=263166

Reviewed by Philippe Normand.

Introduce two new environment variables which control the maximum amount
of buffered data by the downloadbuffer element. This is useful in case
it is necessary to control buffering large files, way larger than the
default downloadbuffer limits. The following new variables are added:

- WPE_SHELL_MEDIA_DISK_CACHE_SIZE_BYTES ~ downloadbuffer max-size-bytes
- WPE_SHELL_MEDIA_DISK_CACHE_SIZE_NSEC ~ downloadbuffer max-size-time

Example HTML file, note esp. the 'preload="auto"' which is mandatory:
```
<video src="./file.mp4" autoplay muted loop preload="auto"></video>
```

Example invocation to cache up to 200 MiB per file and 500s per file:
```
$ WPE_SHELL_MEDIA_DISK_CACHE_SIZE_BYTES=268435456 \
  WPE_SHELL_MEDIA_DISK_CACHE_SIZE_NSEC=500000000000 \
  WEBKIT_GST_USE_PLAYBIN3=1 \
  cog http://example/index.html
```

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleStreamCollectionMessage):

Canonical link: https://commits.webkit.org/269348@main
